File: linked_lists/return_kth_to_last/sol2.py
import sys

sys.path.append('../')

from sll import sll


def kth_to_last(nd, k):
  if not nd:
    return 0
    
  index = kth_to_last(nd.next, k) + 1
  
  if index == k:
    print "kth to the last data is " + str(nd.data)
  
  return index

  
  
if __name__ == "__main__":
  slinked_list = sll.sll()
  value = [1, 2, 3, 1, 3, 4, 5, 1, 2, 3]
  key = 0
  for it in value:
    slinked_list.addLast(key, it)
    key += 1
    
  kth_to_last(slinked_list.head, 2)
  
 
 
 
 
 
 

File: linked_lists/return_kth_to_last/sol2.cpp
#include "sll.h"

#include <iostream>
#include <vector>

int kth_to_last(node* nd, int& target, int& k) {
  if ( !nd ) {
    k = 0;
    return -1;
  }
 
  int res = kth_to_last(nd->next(), target, k);
  ++k;
  if ( k == target ) {
    return nd->data();
  }
  return res;
}


int main() {
  sll slinked_list;
  std::vector<int> value{1, 2, 3, 1, 3, 4, 5, 1, 2, 3};
  int key = 0;
  for (auto& it : value) {
    std::cout << key << std::endl;
    if (key == 0) {
      slinked_list.insertHead(it, key);
    } else {
      slinked_list.insertLast(it, key);
    }
    ++key;
  }
  
  node* nd = slinked_list.head();
  while (nd) {
    std::cout << nd->data() << " ";
    nd = nd->next();
  }
  std::cout << std::endl;
  
  std::cout << "getting result: ";
  int target, k;
  while (std::cin >> target) { 
   int res = kth_to_last(slinked_list.head(), target, k);
   std::cout << "result is " << res << std::endl;
  }
}